storage/innobase/row/row0merge.cc

```diff
@@ -1066,83 +1067,79 @@ static dberr_t row_merge_write_blob_to_tmp_file(
+  blob_file->offset+= (field_len - local_len);
   blob_file->n_rec++;
-  dfield_set_data(field, data, BTR_EXTERN_FIELD_REF_SIZE);
-  dfield_set_ext(field);
   return err;
 }
 
-/** This function is invoked when tuple size is greater than
-innodb_sort_buffer_size. Basically it recreates the tuple
-by writing the blob field to the temporary file.
-@param entry     index fields to be encode the blob
-@param blob_file file to store the blob data
-@param heap      heap to store the blob offset and blob length
-@return tuple which fits into sort_buffer_size */
-static dtuple_t* row_merge_buf_large_tuple(const dtuple_t &entry,
-                                           merge_file_t *blob_file,
-                                           mem_heap_t **heap)
-{
-  if (!*heap)
-    *heap= mem_heap_create(DTUPLE_EST_ALLOC(entry.n_fields));
-
-  dtuple_t *tuple= dtuple_copy(&entry, *heap);
-  for (ulint i= 0; i < tuple->n_fields; i++)
-  {
-    dfield_t *field= &tuple->fields[i];
-    if (dfield_is_null(field) || field->len <= 2000)
-      continue;
-
-    dberr_t err= row_merge_write_blob_to_tmp_file(field, blob_file, heap);
-    if (err != DB_SUCCESS)
-      return nullptr;
-  }
-
-  return tuple;
-}
-
-
-/** Write the field data whose length is more than 2000 bytes
-into blob temporary file and write offset, length into the
-tuple field
-@param entry     index fields to be encode the blob
+/** Variable length field data or fixed length
+character greater than 255 can be stored externally. Instead
+of storing it externally, InnoDB should store it in temporary
+file and write offset, length into the tuple field
+@param fields    index fields to be encode the blob
 @param n_fields  number of fields in the entry
+@param blob_file file to store the blob data
+@param index     index for the tuple to be stored
 @param heap      heap to store the blob offset and blob length
-@param blob_file file to store the blob data */
-static dberr_t row_merge_buf_blob(const mtuple_t *entry, ulint n_fields,
-                                  mem_heap_t **heap, merge_file_t *blob_file)
+@return error code */
+static
+dberr_t row_merge_buf_blob(dfield_t *fields, ulint n_fields,
+                           merge_file_t *blob_file,
+                           const dict_index_t *index,
+                           mem_heap_t **heap)
 {
-
-  if (!*heap)
-    *heap= mem_heap_create(100);
-
-  for (ulint i= 0; i < n_fields; i++)
+  const uint blob_prefix= dict_table_has_atomic_blobs(index->table)
+                          ? 0
+                          : REC_ANTELOPE_MAX_INDEX_COL_LEN;
+  const uint min_local_len = blob_prefix
+                             ? blob_prefix + FIELD_REF_SIZE
+                             : 2 * FIELD_REF_SIZE;
+
+  for (ulint i= index->first_user_field(); i < n_fields; i++)
   {
-    dfield_t *field= &entry->fields[i];
-    if (dfield_is_null(field) || field->len <= 2000)
+    dfield_t *field= &fields[i];
+    if (dfield_is_null(field))
       continue;
```

Comment:
`dtuple_convert_big_rec()` is also checking `index->fields[i].fixed_len`, which we fail to check here.